src/coord/src/coord.rs

```diff
@@ -3459,83 +3484,44 @@ impl<S: Append + 'static> Coordinator<S> {
             was_implicit: session.transaction().is_implicit(),
         };
 
-        // Immediately do tasks that must be serialized in the coordinator.
-        let rx = self
-            .sequence_end_transaction_inner(&mut session, action)
-            .await;
-
-        // We can now wait for responses or errors and do any session/transaction
-        // finalization in a separate task.
-        let conn_id = session.conn_id();
-        task::spawn(
-            || format!("sequence_end_transaction:{conn_id}"),
-            async move {
-                let result = match rx {
-                    // If we have more work to do, do it
-                    Ok(fut) => fut.await,
-                    Err(e) => Err(e),
-                };
-
-                if result.is_err() {
-                    action = EndTransactionAction::Rollback;
-                }
-                session.vars_mut().end_transaction(action);
-
-                match result {
-                    Ok(None) => tx.send(Ok(response), session),
-                    Ok(Some(response)) => tx.send(Ok(response), session),
-                    Err(err) => tx.send(Err(err), session),
-                }
-            },
-        );
-    }
-
-    async fn sequence_end_transaction_inner(
-        &mut self,
-        session: &mut Session,
-        action: EndTransactionAction,
-    ) -> Result<impl Future<Output = Result<Option<ExecuteResponse>, CoordError>>, CoordError> {
-        let txn = self.clear_transaction(session).await;
-
-        let mut write_rx = None;
-
+        let txn = self.clear_transaction(&mut session).await;
         if let EndTransactionAction::Commit = action {
             if let Some(ops) = txn.into_ops() {
-                match ops {
-                    TransactionOps::Writes(mut writes) => {
-                        for WriteOp { id, .. } in &writes {
-                            // Re-verify this id exists.
-                            let _ = self.catalog.try_get_entry(&id).ok_or_else(|| {
-                                CoordError::SqlCatalog(CatalogError::UnknownItem(id.to_string()))
-                            })?;
+                if let TransactionOps::Writes(mut writes) = ops {
+                    for WriteOp { id, .. } in &writes {
+                        // Re-verify this id exists.
+                        if let None = self.catalog.try_get_entry(&id) {
+                            session
```

Comment:
Ah, this is an actual good reason to have a function or other thing to track errors, so we don't have to have an error path that does other arbitrary logic. If we can do a thing that reduces the number of calls to session.vars_mut, I'd prefer that.